### DISCUSSION TITLE
How do I remove workers?

### DISCUSSION THREAD

**Discussion by guilhermeaiolfi** [eModbus/eModbus#200] [Ideas] [answered]
Hi,

I am doing a automatically slave id assigning and it is working fine. The idea is simple, when joining the network, the slave id is pre defined, simething like 127. So I set a worker listening to that id. After the response from the master, telling the real ID for it, I need to remove the worker for the old slave id. 

I register it using: 
```
slave.registerWorker(slave_id, FC, &func);
```

But, there is no ```.deregisterWorker``` yet (as suggested here: https://github.com/eModbus/eModbus/discussions/117). I've tried passing a nullptr to it, but then the response is an Illigal Function Code in the master.

Any other idea how to do it? Or should go ahead and implement the deregisterWorker function?

**Comment by Miq1** [comment]
That would be quite easy to implement. Let me check tomorrow!

**Comment by guilhermeaiolfi** [comment]
Hi,

Just to let you know, I think there should be a way of removing the slave id reference in the map as well (second function), like so:

```cpp
void ModbusServer::unregisterWorker(uint8_t serverID, uint8_t functionCode) {
  auto svmap = workerMap.find(serverID);
  if (svmap != workerMap.end()) {
    svmap->second.erase(functionCode);
    LOG_D("Unregistered worker for %02X/%02X\n", serverID, functionCode);
  }
}

void ModbusServer::unregisterWorker(uint8_t serverID) {
  workerMap.erase(serverID);
  LOG_D("Unregistered worker for %02X\n", serverID);
}
```

For my use case, the first function wasn't enough, I had to use the second one.

**Reply by Miq1** [reply]
I would ask for your patience here, I am about to implement that a bit differently today.

**Comment by Miq1** [comment, marked as answer]
Implemented now - see the [parallel issue](https://github.com/eModbus/eModbus/issues/201) for details.
